src/cpp/session/modules/SessionCopilot.cpp

```diff
@@ -0,0 +1,894 @@
+{
+   s_agentPid = -1;
+
+   if (status != 0)
+   {
+      ELOG("Agent exited with status {}.", status);
+   }
+}
+
+} // end namespace agent
+
+void stopAgent()
+{
+   if (s_agentPid == -1)
+   {
+      DLOG("No agent running; nothing to do.");
+      return;
+   }
+
+   Error error = core::system::terminateProcess(s_agentPid);
+   if (error)
+      LOG_ERROR(error);
+}
+
+bool startAgent()
+{
+   Error error;
+
+   FilePath agentPath = copilotAgentPath();
+   if (!agentPath.exists())
+   {
+      ELOG("Copilot agent not installed; cannot start agent.");
+      return false;
+   }
+
+   // Create environment for agent process
+   core::system::Options environment;
+   core::system::environment(&environment);
+
+   // For Desktop builds of RStudio, use the version of node embedded in Electron.
+   FilePath nodePath;
+   if (session::options().programMode() == kSessionProgramModeDesktop)
+   {
+       std::string exePath = core::system::getenv("RSTUDIO_DESKTOP_EXE");
+       if (!exePath.empty() && FilePath(exePath).exists())
+       {
+          nodePath = FilePath(exePath);
+          environment.push_back(std::make_pair("ELECTRON_RUN_AS_NODE", "1"));
+       }
+   }
+
+   // Otherwise, use node from the PATH
+   // TODO: We'll need to bundle a version of node with RStudio Server.
+   // Quarto 1.4 will ship with 'deno', which will be able to run regular 'node' applications,
+   // so maybe we can use that?
+   if (!nodePath.exists())
+   {
+      error = core::system::findProgramOnPath("node", &nodePath);
+      if (error)
+      {
+         LOG_ERROR(error);
+         return false;
+      }
+   }
+
+   // Set up process callbacks
+   core::system::ProcessCallbacks callbacks;
+   callbacks.onStarted = &agent::onStarted;
+   callbacks.onContinue = &agent::onContinue;
+   callbacks.onStdout = &agent::onStdout;
+   callbacks.onStderr = &agent::onStderr;
+   callbacks.onExit = &agent::onExit;
+
+   // Set up process options
+   core::system::ProcessOptions options;
+   options.environment = environment;
+   options.allowParentSuspend = true;
+   options.exitWithParent = true;
+   options.callbacksRequireMainThread = true; // TODO: It'd be nice to drop this requirement!
+   options.workingDir = agentPath.getParent();
+
+   error = module_context::processSupervisor().runProgram(
+            nodePath.getAbsolutePath(),
+            { agentPath.getAbsolutePath() },
+            options,
+            callbacks);
+
+   if (error)
+   {
+      LOG_ERROR(error);
+      return false;
+   }
+
+   // Wait for the process to start.
+   //
+   // TODO: This is kind of a hack. We should probably instead use something like a
+   // status flag that tracks if the agent is stopped, starting, or already running.
+   //
+   // We include this because all requests will fail if we haven't yet called
+   // initialized, so maybe the right approach is to have some sort of 'ensureInitialized'
+   // method?
+   waitFor([]() { return s_agentPid != -1; });
```

Comment:
Let's keep this synchronous for now and revisit if it turns out to be a problem down the line.